apps/wallet/src/background/keyring/VaultStorage.test.ts

```diff
@@ -0,0 +1,318 @@
+                },
+            });
+            expect(getRandomEntropySpy).toHaveBeenCalledOnce();
+        });
+
+        it('uses the provided entropy and creates the vault', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({});
+            const storageSet = vi
+                .spyOn(Browser.storage.local, 'set')
+                .mockResolvedValue();
+            const getRandomEntropySpy = vi
+                .spyOn(Bip39, 'getRandomEntropy')
+                .mockReturnValue(new Uint8Array(32));
+            await vaultStorage.create(
+                '12345',
+                '842a27e29319123892f9ba8d9991c525'
+            );
+            expect(storageSet).toHaveBeenCalledOnce();
+            expect(storageSet.mock.calls[0][0]).toMatchObject({
+                vault: {
+                    v: 2,
+                    data: expect.stringContaining('data'),
+                },
+            });
+            expect(getRandomEntropySpy).not.toHaveBeenCalled();
+        });
+    });
+
+    describe('unlock', () => {
+        it('throws if wallet is not initialized', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({});
+            await expect(vaultStorage.unlock('12345')).rejects.toThrow(
+                'Wallet is not initialized. Create a new one first.'
+            );
+        });
+
+        it('throws if password is wrong', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({
+                vault: testVault.encrypted.v2,
+            });
+            await expect(vaultStorage.unlock('wrong password')).rejects.toThrow(
+                'Incorrect password'
+            );
+        });
+
+        it('unlocks the vault', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({
+                vault: testVault.encrypted.v2,
+            });
+            await vaultStorage.unlock(testVault.password);
+            expect(vaultStorage.getMnemonic()).toBe(testVault.mnemonic);
+        });
+
+        it('unlocks the vault and updates session storage', async () => {
+            const vaultStorage = new VaultStorage();
+            vi.spyOn(Browser.storage.local, 'get').mockResolvedValue({
+                vault: testVault1.encrypted.v2,
+            });
+            await vaultStorage.unlock(testVault.password);
+            expect(vaultStorage.getMnemonic()).toBe(testVault.mnemonic);
+            const sessionStorageSet = vi.mocked(SESSION_STORAGE!.set);
+            expect(sessionStorageSet).toBeCalledTimes(2);
+            expect(sessionStorageSet.mock.calls[0][0]).toMatchObject({
+                '244e4b24e667ebf': expect.stringMatching(/.+/),
```

Comment:
These keys seem like magic values, probably should be imported from the module. Also you can use `expect.any(String)`

Reply:
I think `expect.stringMatching(/.+/)` catches an extra case that the string is not empty